### PR TITLE
speedscope format: make units conform to schema

### PR DESCRIPTION
**Pull request by goxberry**
### Description of the Change

This pull request replaces the (Unicode) string `"\u03bcs"` in the speedscope `Units`
enumeration with the string `"microseconds"`, so that the `"units"`
field of Speedscope-formatted Austin CPU profiles conforms to the
Speedscope JSON schema defined in
https://github.com/jlfwong/speedscope/blob/main/src/lib/file-format-spec.ts.
Predecessor commits in this repo emit Speedscope-formatted JSON that
does not conform to the aforementioned JSON schema, and as a result, when Speedscope renders a timeline view of the Austin CPU profile, the timeline has no units. By making this string replacement, Speedscope renders a timeline view of the Austin CPU profile using the correct units.

### Alternate Designs

The only other conforming strings for the `units` field of Speedscope-formatted JSON files, as described in https://github.com/jlfwong/speedscope/blob/main/src/lib/file-format-spec.ts, are:

* `'none'`
* `'nanoseconds'`
* `'milliseconds'`
* `'seconds'`
* `'bytes'`

Since the units in question are indeed microseconds for a CPU profile, none of these alternatives is appropriate. The only other sensible choice would be `'none'`, but then that choice would still result in timeline views without time units, which is what this pull request is trying to fix.

### Regressions

> What are the possible side-effects or negative impacts of the code change?

Since the modification made is to the value of one field in an enumeration, barring Python syntax errors, there should be no negative impacts of this code change. Python syntax errors aside, at worst, the `units` field of Speedscope-formatted Austin profiles still fails to conform to the Speedscope JSON spec, in which case the rendered profile will still look the same as before. If the change succeeds as intended, the rendered profile will have time units on the horizontal axis of its timeline view.

> Would any user of austin be forced to adapt their scripts or code if they wanted
to upgrade to the new version?

Users would have to run `austin2speedscope` again to convert profile data from Austin format to Speedscope format; otherwise, I would expect most users would not have to adapt scripts or code.

If users postprocess the Speedscope-formatted Austin profile to correct the issue with the `units` field addressed in this pull request, users would be forced to remove that postprocessing step. I would expect such a change to require minimal code adaptation (no more than 10 lines of Python, possibly even a single line of Python). I view this change as positive, because it demands less of users in the future.

### Verification Process

> What process did you follow to verify that your change has the desired effects?
> 
> - How did you verify that all new functionality works as expected?
> - How did you verify that all changed functionality works as expected?
> - How did you verify that the change has not introduced any regressions?
> - How did you verify that the change has not introduced any memory leaks?

In my local copy of the repo, I ran:

```bash
poetry config virtualenvs.in-project true
poetry shell
poetry install
austin2speedscope np_c_function.aprof np_c_function_austin.json  # profile of a script I had lying around
python3 -m json.tool < np_c_function_austin.json > np_c_function_austin_pretty.json
```
and examined the prettified JSON before and after the changes made in this pull request. Visualized results are also attached.

#### Before -- note the lack of units on the timeline at top of screenshot
AustinSpeedscopeBefore

#### After -- note that units are now on the timeline at top of screenshot (Speedscope uses milliseconds)
AustinSpeedscopeAfter

### Additional details (motivation, more background)

In predecessors to this commit, the `austin2speedscope` command line
tool would convert an Austin (CPU time) profile to speedscope format
such that the `profiles[i].units` field for any valid index of `i`
(e.g., 0) would contain the value `"\u03bcs"`, i.e., the UTF-8 code
point U+03BC -- displayed as the Greek letter "mu" in compatible
editors and terminals -- followed by the lowercase letter "s". While
this choice undoubtedly looks cool and displays as if one were
abbreviating the unit "microseconds" as in scientific and engineering
texts, this choice does not conform to the Speedscope JSON schema as
of upstream Speedscope commit
`e4040538376bbaa81f3746f11530717a1f2544b` (see
https://github.com/jlfwong/speedscope/commit/e4040538376bbaa81f3746f11530717a1f2544b1#diff-5bd093fe1cb8e10b3f1ae3621ec86c72191d17a90d73e4145518766b2d6d1ce3R117,
circa 2018-August), which has not since been changed as of this
writing. (For reference, as of this writing, the most recent upstream
commit is `e37f6fa7c38c110205e22081560b99cb89ce885e` and the relevant
link to the schema definition is at
https://github.com/jlfwong/speedscope/blob/e37f6fa7c38c110205e22081560b99cb89ce885e/src/lib/file-format-spec.ts#L117.)

Replacing the string `"\u03bcs"` with the string `"microseconds"`
does, however, conform to the Speedscope JSON schema, and results in
displaying the reformatted Austin profile in Speedscope's timeline
view with the proper units. Without this change, the Speedscope
timeline view omits units altogether. A viewer then has to simply know
that the horiztonal axis of the Speedscope icicle chart has units of
microseconds, even though those units are omitted from the graphic,
which makes the resulting graphic harder to use in presentations to
other developers and to non-technical management.

This pull request makes that replacement.

### CODE REVIEW

**Comment by goxberry** [comment]
I'd also suggest rewriting the project `CONTRIBUTING.md` to reflect that this project is written in Python, not C, and would suggest outlining how contributors can set up a virtual environment to use for development and testing. Some of the pull request template (e.g., the part about memory leaks), doesn't obviously seem to apply to Python code, though I do appreciate the questions regarding effects of contributions on users and developers. I also appreciate the thought and effort you've put into the tooling in this project and in the Austin project as well; it's a really cool tool I'm trying to get approved for use by my employer.

As for testing, as best as I can tell, the Speedscope formatter test in https://github.com/P403n1x87/austin-python/blob/main/test/format/test_speedscope.py doesn't test very much for adherence of the JSON output to the Speedscope schema. The change I made is simple enough that it didn't seem worth going through the effort to add better coverage for JSON schema conformance to the existing tests, but doing so might be a good issue for new contributors.

**Comment by P403n1x87** [comment]
Wow, that's how all PR descriptions should look like! :muscle: 

Many thanks for taking the time to make the change and write such a detailed description for it! The `austin2speedscope` should definitely adhere to the JSON schema to make the best of the Speedscope.app features.

> As for testing, as best as I can tell, the Speedscope formatter test in https://github.com/P403n1x87/austin-python/blob/main/test/format/test_speedscope.py doesn't test very much for adherence of the JSON output to the Speedscope schema

Indeed testing of the `austin2speedscope` too is not great at the moment. Perhaps we could switch to a snapshot test like we currently do for `pprof`

https://github.com/P403n1x87/austin-python/blob/c5693664f0ae20191ce67b208b81bd2010d23bfa/test/format/test_pprof.py#L35-L50

Let me know if you would like to give it a go, otherwise we can enhance the test in a follow-up PR.

----

> I'd also suggest rewriting the project `CONTRIBUTING.md` to reflect that this project is written in Python, not C

:tada: Congratulations on discovering that this file was shamelessly copied over from the [Austin repository](https://github.com/P403n1x87/austin/blob/0354d4b22f0b7bbad8da11c9f0941ce945cff220/CONTRIBUTING.md) :grin: . Seriously, thanks for the great suggestions on how to improve this and the repository templates. I'll make sure to push some improvements as soon as I can! :+1:

**Comment by goxberry** [comment]
Based on what I could figure out from the codebase, I've added more comprehensive collections of assertions to the existing two unit tests of the Speedscope format writer in [`72a2e001`](https://github.com/P403n1x87/austin-python/pull/7/commits/72a2e0010ca287c0366fe80536ef5fe98e56a7dd).

I also started adding a test along the lines of the `pprof` test you suggested. However, doing so uncovered a few issues, some of which I have fixed:

* If the Austin profiling mode is `"wall"`, then https://github.com/P403n1x87/austin-python/blob/abffb7577695d093661fe69369f0f1e45b9b7779/austin/format/speedscope.py#L136-L137 forwards the data as if it corresponded to the `"cpu"` Austin profiling mode. This bug is fixed in [`1c047179`](https://github.com/P403n1x87/austin-python/pull/7/commits/1c04717925e79006ce5a7aa5c31f139a5bd500f3).
* The docstring of `austin.format.Speedscope.dump` in https://github.com/P403n1x87/austin-python/blob/c5693664f0ae20191ce67b208b81bd2010d23bfa/austin/format/speedscope.py#L164-L170 appears to be incorrect because it refers to `pprof` and binary streams, even though it serializes JSON to text. This docstring is corrected in [`7c3ccd49`](https://github.com/P403n1x87/austin-python/pull/7/commits/7c3ccd49e62a39e76225ba9dde8ee5ca844590c3).
* The `DATA_FILE` string in https://github.com/P403n1x87/austin-python/blob/c5693664f0ae20191ce67b208b81bd2010d23bfa/test/__init__.py#L1-L3 uses a relative path, which is brittle because running `pytest /path/to/test_speedscope.py` within the project directory tree will fail unless the current working directory is the project root. I replaced the relative path with an absolute one computed from `os.abspath(os.dirname(__file__))` via [`409a18c8`](https://github.com/P403n1x87/austin-python/pull/7/commits/409a18c828a5bca314fd88314893316b2fd0417d). There are relative paths used in other tests, like the `pprof` test, which are brittle for similar reasons to the previous point; however, I elected to leave those other paths alone because those files are not Speedscope-related.
* The profiling data in `test/data/austin.json` does not correspond to the profiling data in `test/data/austin.out`. For example, the process IDs are different, the thread IDs are different, and file paths within `austin.out` don't match those in `austin.json` (e.g., `austin.out` has paths rooted at `/usr/lib/python3.8`, but `austin.json` has paths rooted at `/usr/lib/python3.6`). Rather than blindly overwrite the data in `austin.json` using the `austin.format.speedscope.Speedscope` implementation in my branch, I used some quick sequences of Unix text processing commands to extract data from `austin.out` so I could hardcode some tests of the conversion to Speedscope format in a new `test_speedscope_wall_metrics_only` test.
* After I wrote enough hardcoded tests to be confident that the `austin2speedscope` script would work correctly, I also updated the contents of `austin.json` using `austin.out` via [`6b1554cc`](https://github.com/P403n1x87/austin-python/pull/7/commits/6b1554cccbaf25b438378d063b02292f9ff4b3ac).

The hardcoded tests in `test_speedscope_wall_metrics_only` don't test the `profiles[i].samples` arrays for `i = 0` or `i = 1`. Despite employing several command-line tricks to facilitate manual inspection of data in `austin.out`, attempting to reconstruct manually stack frames for each of 72 samples seemed like an error-prone process.

I left examples of the commands I used for manual data extraction in the comments, in case the contents of `austin.out` are updated. Manually determining stack frame indices is probably too error-prone as soon as the number of elements in `shared.frames` exceeds 15. (The structure of the data in `austin.out` was pretty fortuitous for that task because the first 7 or 8 function calls in `austin.out` are distinct, leaving me to find 3-4 indices in the remainder of the data.) But manually determining the frames in `shared.frames` without determining their exact positions in that array should be more scalable.

File: austin/format/speedscope.py
Comment (P403n1x87) [inline austin/format/speedscope.py:137]: Ah! Well spotted! :eyes:

File: austin/format/speedscope.py
Comment (P403n1x87) [inline austin/format/speedscope.py:165]: More brainless copy-pasta from me :disappointed:

File: test/__init__.py
Comment (P403n1x87) [inline test/__init__.py:6]: :+1: :+1:

File: test/data/austin.json
Comment (P403n1x87) [inline test/data/austin.json:1]: This file was originally generated by calling `austin2speedscope` on `test/data/austin.out` with

~~~ console
poetry run austin2speedscope test/data/austin.out test/data/austin.json
~~~

 and was not updated when the latter was changed. I would have expected simply regenerating this file and making sure that it looks reasonable on Speedscope.app would have done the trick. This is what the pprof test does anyway, given the complexity of these file formats, even for relatively simple examples.

Aside: there should probably be a task in the `noxfile` to generate these snapshots when `austin.out` is changed. I'll look into this.

File: test/format/test_speedscope.py
Comment (P403n1x87) [inline test/format/test_speedscope.py:241]: :+1:  This is great but I wonder if this logic could be simplified once the `austin.json` file is regenerated from the current `austin.out`.

**Review by P403n1x87 — COMMENTED** [review]
:100: This is all great stuff. Thanks for all the efforts to improve this area of the library! :bow:

I have left a few comments, but otherwise LGTM overall.

File: test/data/austin.json
Comment (goxberry) [reply test/data/austin.json:1]: As far as I understand it, I did something equivalent to regenerate `austin.json` from `austin.out` in [`6b1554cc`](https://github.com/P403n1x87/austin-python/pull/7/commits/6b1554cccbaf25b438378d063b02292f9ff4b3ac).:

```bash
poetry shell
austin2speedscope test/data/austin.out test/data/austin.json
```

File: test/format/test_speedscope.py
Comment (goxberry) [reply test/format/test_speedscope.py:241]: As far as I understand it, I've already committed a newer version of `austin.json` that shouldn't require regenerating again for this PR.

Had I regenerated the JSON before adding the hardcoded tests, I probably wouldn't have caught the error re: `cpu` vs `wall` metrics because the only discrepancies found between the prior `austin.json` and the JSON generated by the code (sans `cpu` vs `wall` fix) would have been due to the change in the `unit` field (re: microseconds), file paths, process IDs, and thread IDs. As a diagnostic, comparing against the contents of `austin.json` will tell you whether the JSON output from the code has changed for regression testing, but if the `austin.json` file itself has errors in it, then "passing" that test may be a false negative.

In the context of this project, I don't have a particularly a strong opinion on adding correctness tests to `test_speedscope_wall_metrics_only`, and I'm happy to remove those tests, because even with scripts to lessen the effort costs of data entry, it's still a substantial amount of work to extract the correct values.

Comment (goxberry) [reply test/format/test_speedscope.py]: Removed in [`ac5da98f`](https://github.com/P403n1x87/austin-python/pull/7/commits/ac5da98ff3cbff3445b89e66e39ff1d4f07af3ef).

File: austin/format/speedscope.py
Comment (goxberry) [reply austin/format/speedscope.py:165]: No worries; I've definitely done that myself!

File: test/data/austin.json
Comment (goxberry) [reply test/data/austin.json:1]: If there's anything else I should do re: the `austin.json`, let me know and I'll be happy to make the necessary changes.

Comment (P403n1x87) [reply test/data/austin.json:1]: Ah, my bad! I've missed that the first time around! I now realise that `test_speedscope_wall_metrics_only` tests against the updated snapshot, so as long as the new `austin.json` looks reasonable we should be good! 👍 

Just as a nit, we could regenerate `austin.json` with e.g. `indent=2` so that it is more readable and easier to use to spot potential regressions in the future. But I leave this up to you.

**Review by P403n1x87 — APPROVED** [review]

File: test/data/austin.json
Comment (goxberry) [reply test/data/austin.json]: I agree with your suggestion re: indentation.

In [`9a56b7f8`](https://github.com/P403n1x87/austin-python/pull/7/commits/9a56b7f8af1f9ad1e088c02238ae000da74df46b) I've updated the `test_speedscope_wall_metrics_only` test to use `json.load` and `json.dumps`, which makes the string comparison in the assertion insensitive to whitespace in `austin.json`. This change made it easier to change the indentation in `austin.json` without breaking existing tests.

I then made the change to add indentation to `austin.json` -- as you suggested -- in [`43e70bd5`](https://github.com/P403n1x87/austin-python/pull/7/commits/43e70bd5bf80f9744e7ec701d2e11030795b42f5).

**Comment by goxberry** [comment]
@P403n1x87 One quick question, if you'll indulge me: does Austin collect timing data on time spent in C extensions of Python packages? I've noticed that other stack sampling Python profilers, like `pyinstrument`, have a limited ability, if any, to collect such data.

**Comment by P403n1x87** [comment]
> does Austin collect timing data on time spent in C extensions of Python packages?

I have recently merged some changes into the `devel` branch that allow compiling a variant of Austin, named `austinp`, that can sample native stacks on Linux (in fact, it can also sample down to the Linux kernel level). A nightly build is available from the development release https://github.com/P403n1x87/austin/releases/tag/dev. It works like Austin, except that by default you also get the native stacks (whether it's C/C++/Cython/Rust/..., as long as the compiled binaries export symbols). See https://github.com/P403n1x87/austin/tree/devel#native-frame-stack for more details on how to get the symbols resolved. This feature will land into the main branch with the next release.

**Comment by goxberry** [comment]
That's awesome! I look forward to seeing it!

File: test/format/test_speedscope.py
Comment (P403n1x87) [inline test/format/test_speedscope.py:236]: I think we should be able to keep this as
```suggestion
            assert text_stream.getvalue() == sprof.read()
```
if we change line 224 above into
~~~ python
speedscope = Speedscope("austin.out", mode, indent=2)
~~~

**Review by P403n1x87 — COMMENTED** [review]
This is great! Thanks for all the work! Just a nit, otherwise LGTM!

File: test/format/test_speedscope.py
Comment (goxberry) [reply test/format/test_speedscope.py]: Done in [`f09f13ae`](https://github.com/P403n1x87/austin-python/pull/7/commits/f09f13ae93e3ad2c0b3e71189438f7f14a6dff15).

**Review by P403n1x87 — APPROVED** [review]
:+1: :+1:

**Pull request merged**
